### PR TITLE
Application Charm Config Read and Write Uses Generation Argument

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch ensures that generational argument passed through to the application API server are accommodated in state methods for getting and setting charm configuration.

### Main Feature
- `export JUJU_DEV_FEATURE_FLAGS=generations`
- Bootstrap
- `juju deploy redis`
- `juju config redis password=current`
- `juju add-generation`
- `juju config redis password=next`
- `juju config redis password --generation current` shows "current".
- `juju config redis password --generation next` shows "next".

### Bonus - Current Watch Behaviour Not Broken By Generational Config
- In a separate terminal: `juju debug-log --include redis/0`
- `juju config redis password=current-new --generation current` and observe the config-changed hook fire.
- `juju config redis password=next-new --generation next` and observe that the config-changed hook does not fire.

## Documentation changes

Eventual documentation for the new feature.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Write Generational Application Config](https://trello.com/c/PXYEjCN9/1079-write-generational-application-config)

File: apiserver/facades/client/application/application.go [hunk 1]
```diff
@@ -678,9 +679,11 @@ func checkMachinePlacement(backend Backend, args params.ApplicationDeploy) error
 	return nil
 }
 
-// ApplicationSetSettingsStrings updates the settings for the given application,
+// applicationSetSettingsStrings updates the settings for the given application,
```
Comment (manadart) [inline apiserver/facades/client/application/application.go]: This didn't need exporting. Drive-by to make it private for congruence with the other methods.

File: state/application.go [hunk 2]
```diff
@@ -120,6 +121,24 @@ func applicationCharmConfigKey(appName string, curl *charm.URL) string {
 	return fmt.Sprintf("a#%s#%s", appName, curl)
 }
 
+// charmConfigKeyGeneration returns the charm-version-specific settings
+// collection key and possibly a fallback, for the application based on the
+// input generation. I
```
Comment (achilleasa) [inline state/application.go]: Accidental line-break?

File: state/settings.go [hunk 3]
```diff
@@ -315,6 +315,27 @@ func (st *State) ReadSettings(collection, key string) (*Settings, error) {
 	return readSettings(st.db(), collection, key)
 }
 
+// readSettingsWithFallback attempts to retrieve settings first for the
+// requested key, then if not found, a non-empty fallback key.
+// If the fallback is used, the settings are created for the requested key.
+func readSettingsWithFallback(db Database, collection, requestKey, fallbackKey string) (*Settings, error) {
```
Comment (achilleasa) [inline state/settings.go]: The function name seems a bit misleading as the the function can potentially create the settings by copying the fallback ones.

**Comment by mitechie** [comment]
QA ok

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
This one again.
`FAIL: apiclient_test.go:792: apiclientSuite.TestOpenTimeoutAffectsDial`
$$merge$$

**Comment by manadart** [comment]
`mongod failed to listen on port 34364`
$$merge$$
